sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -161,23 +161,49 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
           bucketSpec,
           Some(partitionSchema))
 
+        val catalogTable = metastoreRelation.catalogTable
         val logicalRelation = cached.getOrElse {
           val sizeInBytes =
             metastoreRelation.stats(sparkSession.sessionState.conf).sizeInBytes.toLong
           val fileIndex = {
-            val index = new CatalogFileIndex(
-              sparkSession, metastoreRelation.catalogTable, sizeInBytes)
+            val index = new CatalogFileIndex(sparkSession, catalogTable, sizeInBytes)
             if (lazyPruningEnabled) {
               index
             } else {
               index.filterPartitions(Nil)  // materialize all the partitions in memory
             }
           }
           val partitionSchemaColumnNames = partitionSchema.map(_.name.toLowerCase).toSet
-          val dataSchema =
-            StructType(metastoreSchema
+          val filteredMetastoreSchema = StructType(metastoreSchema
               .filterNot(field => partitionSchemaColumnNames.contains(field.name.toLowerCase)))
 
+          val inferenceMode = sparkSession.sessionState.conf.schemaInferenceMode
+          val dataSchema = if (inferenceMode != "NEVER_INFER" &&
+              !catalogTable.schemaFromTableProps) {
+            val fileStatuses = fileIndex.listFiles(Nil).flatMap(_.files)
+            val inferred = defaultSource.inferSchema(sparkSession, options, fileStatuses)
+            val merged = if (fileType.equals("parquet")) {
+              inferred.map(ParquetFileFormat.mergeMetastoreParquetSchema(metastoreSchema, _))
+            } else {
+              inferred
```

Comment:
I took this from how the schema was inferred in HiveMetastoreCatalog prior to 2.1.0. Only ParquetFileFormat has a merge method.